data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/test/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/auth/Office365AuthenticationProviderTest.java

```diff
@@ -104,6 +131,141 @@ void testRenewCredentialsWithInvalidResponse() {
+        String token = authProvider.getAccessToken();
+
+        assertEquals("testAccessToken", token);
+        verify(clientIdVariable).refreshAndRetrieveValue();
+        verify(clientSecretVariable).refreshAndRetrieveValue();
+    }
+
+    @Test
+    void testConcurrentAccessTokenInitialization() throws InterruptedException, NoSuchFieldException, IllegalAccessException {
+        mockSuccessfulTokenResponse();
+
+        // Ensure access token is null initially
+        ReflectivelySetField.setField(Office365AuthenticationProvider.class, authProvider, "accessToken", null);
+
+        int threadCount = 10;
+        ExecutorService executor = Executors.newFixedThreadPool(threadCount);
+        CountDownLatch latch = new CountDownLatch(threadCount);
+        String[] results = new String[threadCount];
+
+        // Start multiple threads trying to get access token simultaneously
+        for (int i = 0; i < threadCount; i++) {
+            final int index = i;
+            executor.submit(() -> {
+                try {
+                    results[index] = authProvider.getAccessToken();
+                } finally {
+                    latch.countDown();
+                }
+            });
+        }
+
+        // Wait for all threads to complete
+        assertTrue(latch.await(5, TimeUnit.SECONDS));
+        executor.shutdown();
+
+        // Verify all threads got the same token
+        for (String result : results) {
+            assertEquals("testAccessToken", result);
+        }
+
+        // Verify initCredentials was called only once due to double-checked locking
+        verify(clientIdVariable, times(1)).refreshAndRetrieveValue();
+        verify(clientSecretVariable, times(1)).refreshAndRetrieveValue();
+    }
+
+    @Test
+    void testIsCredentialsInitializedReturnsFalseInitially() {
+        // Credentials should not be initialized initially
+        assertFalse(authProvider.isCredentialsInitialized());
+    }
+
+    @Test
+    void testIsCredentialsInitializedReturnsTrueAfterSuccessfulInit() {
+        mockSuccessfulTokenResponse();
+
+        // Initialize credentials
+        authProvider.initCredentials();
+
+        // Should return true after successful initialization
+        assertTrue(authProvider.isCredentialsInitialized());
+    }
+
+    @Test
+    void testInitCredentialsBlocksUntilSuccessfulInitialization() throws InterruptedException {
+        // Mock multiple failures followed by success
+        when(restTemplate.postForEntity(anyString(), any(HttpEntity.class), eq(Map.class)))
+                .thenThrow(new HttpClientErrorException(HttpStatus.BAD_REQUEST))
+                .thenThrow(new HttpServerErrorException(HttpStatus.INTERNAL_SERVER_ERROR))
+                .thenReturn(createSuccessfulTokenResponse());
+
+        // Use a separate thread to test the retry behavior
+        AtomicReference<Exception> thrownException = new AtomicReference<>();
+        
+        Thread initThread = new Thread(() -> {
```

Comment:
Rather than a separate thread, can you use Awaitility to wait for a condition to indicate it retried? I'm guessing this would be the `verify(restTemplate...)` that you have one line 233.